Add RequestStarted to GlobalStatusCode enum

nemanja-boric-sociomantic opened the pull request:
This PR adds support for moving the initial "request supported" handshake into the swarm. The request should be adapted so they don't send the request implementation defined `Started` code, but to just send `RequestStarted` when they are ready to start handling the request. Then they can send `Error` message (which is request defined) any time to the client, which should also expect it and abort everything with the `node_error`.

This is to allow request handlers to be `nothrow`, notifying the client about the error that just happened, and which can happen any time.

Node's connection handler then can be enhanced with the nothrow enforcement, but that needs to be done in the next major.

gavin-norman-sociomantic reviewed (commented):
A high-level comment that applies to all commits:

I'd suggest that we rename things a bit:
* `GlobalStatusCode` -> `SupportedCode`.
* `RequestStarted` -> `RequestSupported`.
* `handleRequestStatusCodes` -> `handleSupportedCodes`

The current sole function of this code is to indicate supported / not, so unless there are any ideas for it being used to carry any other messages, we may as well make it clear what it represents.

nemanja-boric-sociomantic commented:
Updated and rebased to v4.x.x. I'm not sure what to do about `GlobalStatusCode` enum. I feel `deprecating` it here would bring more harm than good, as it would propagate to all current requests.

nemanja-boric-sociomantic commented:
(and it's not an internal request detail, like `nextEvent`, but actually a part of the node/client protocol.)

gavin-norman-sociomantic commented on src/swarm/neo/request/Command.d:
```diff
@@ -61,6 +61,11 @@ public alias byte StatusCode;
+    /// Both request code specified in the transmitted Command struct and the
+    /// associated request version are supported and the request will start
+    /// on this connection.
+    RequestSupported = -3,
```
The commit message of "Add RequestStarted to GlobalStatusCode enum" is outdated.

gavin-norman-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
+
+    ***************************************************************************/
+
+    private static bool handleSupportedCodes ( StatusCode status,
```
As we no longer intend to allow the set of codes to be "extended" per request type, this function should just accept `SupportedStatus` directly.

gavin-norman-sociomantic commented on src/swarm/neo/request/Command.d:
```diff
@@ -61,6 +61,11 @@ public alias byte StatusCode;
 
 public enum GlobalStatusCode : StatusCode
```
Maybe add `alias GlobalStatusCode SupportedStatus;`.

gavin-norman-sociomantic replied:
Oh, or would renaming the enum and adding a deprecated alias do the job? (As long as we change all usages inside swarm, we shouldn't get deprecation warnings.)

nemanja-boric-sociomantic replied:
No, sadly, this is used outside the swarm (the node handlers are sending these codes to the clients).

gavin-norman-sociomantic replied:
Right, that would mean that when we update *proto to use swarm v4.5.0, we'd get deprecation warnings that we'd then fix. Or?

nemanja-boric-sociomantic replied:
Right, that would work, it's just renaming in this case.

gavin-norman-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
+            remote_address = address of node which returned this status code
+
+        Returns:
+            false if `status` was a global error code and the user notified;
```
"global error code" is outdated terminology.

gavin-norman-sociomantic commented on integrationtest/neo/client/request/internal/GetAll.d:
```diff
@@ -275,32 +275,11 @@ private scope class GetAllImpl
     {
         auto getall_status = cast(RequestStatusCode)status;
```
This variable is a bit deceptive now, as there are no special GetAll status codes.

gavin-norman-sociomantic commented on integrationtest/neo/client/request/internal/GetAll.d:
```diff
@@ -393,21 +373,34 @@ private struct Reader
-            }
-        );
-        this.conn.flush();
+        if (!error)
```
Could this be moved inside `case End:`?

gavin-norman-sociomantic commented on integrationtest/neo/node/request/GetAll.d:
```diff
@@ -38,15 +38,23 @@ public void handle ( Object shared_resources, RequestOnConn connection,
+                {
+                    payload.addConstant(GlobalStatusCode.RequestSupported);
+                }
+            );
```
This ought to be flushed.

gavin-norman-sociomantic replied:
(All the more reason for https://github.com/sociomantic-tsunami/swarm/issues/168.)

gavin-norman-sociomantic commented on integrationtest/neo/node/request/GetAll.d:
```diff
@@ -218,33 +226,36 @@ private scope class GetAllImpl_v0
+            this.writer.fiber.start();
+            this.request_event_dispatcher.eventLoop(this.conn);
+        }
+        catch (Exception e)
```
I really wonder if this would after all be worth handling at a higher level. The only reason I can think of to not do that is that the exact code to send is defined per request (atm).

nemanja-boric-sociomantic replied:
It's a tricky situation. The reason I didn't want it there, is that perhaps different requests may define different ways how the errors are reported to the client (and it requires change of the request definitions to expect it anytime). I guess I could add the separate method that runs the handler, accepts the error code (or use the swarm defined, in which case we'll have to add one), but then it's the question how should we plug this with the existing request.

Perhaps we should just add the request code and send it, as, if the client is not aware of it, it will shut down with the protocol error, which is still better than the current situation?

gavin-norman-sociomantic replied:
I suppose we can leave it down to each request for now, then see what common pattern develops. (I suspect all requests will end up doing the same, but we may be surprised.)

nemanja-boric-sociomantic committed:
Rename GlobalStatusCode enum to SupportedStatus

This enum will now not be request-extensible, as it will be a part of
the fixed request handshake process where the node just checks if the
request and the request version are supported, and responds to the
client. Hence, we rename it to reflect that it contains only info about
request support.

nemanja-boric-sociomantic committed:
Add RequestSupported value to GlobalStatusCode enum

This value will be used by the requests in the future and it will
mean that the request is supported and just had started on the
node.

nemanja-boric-sociomantic committed:
Add RequestCore.handleSupportedCodes helper method

Since the GlobalStatusCode handling is now deprecated in favour of the
simple request support check, the new helper method is added which
client should use to check if the request and the request versions are
supported.

nemanja-boric-sociomantic committed:
Adapt GetAll to expect the Error message anytime

This adapts GetAll code to use RequestStarted/Request*NotSupported
mechanism to inform client if the request is started or rejected because
it's not supported, and to allow node to send the (now added) Error
message any time it finds suitable (for example, any time the request
ends because of the exception on the node's side).

nemanja-boric-sociomantic commented:
Updated.

gavin-norman-sociomantic commented on integrationtest/neo/client/request/internal/GetAll.d:
```diff
@@ -273,7 +274,7 @@ private scope class GetAllImpl
     private bool handleStatusCode ( ubyte status )
     {
-        auto request_status = cast(RequestStatusCode)status;
+        auto request_status = cast(SupportedStatus)status;
```
This looks unrelated to this commit?

nemanja-boric-sociomantic replied:
Hm, I wanted to get rid of all the deprecations, but I should actually get rid of them in a good way - actually using the new mechanism.

nemanja-boric-sociomantic committed:
Flush after sending the RequestVersionNotSupported code in test request

nemanja-boric-sociomantic commented:
Updated.

gavin-norman-sociomantic commented:
Hmm release notes? The deprecation could be mentioned, say.

nemanja-boric-sociomantic commented:
Oh, damn, that slipped me!

gavin-norman-sociomantic commented:
Yeah, and new feature.

nemanja-boric-sociomantic commented:
Updated with :police_car:

gavin-norman-sociomantic commented on src/swarm/neo/client/mixins/RequestCore.d:
```diff
@@ -375,6 +375,57 @@ public template RequestCore ( RequestType request_type_, ubyte request_code,
+
+    ***************************************************************************/
+
+    private static bool handleSupportedCodes ( SupportedStatus status,
```
Actually, is there any reason to not deprecate `handleGlobalStatusCodes`?

nemanja-boric-sociomantic replied:
Yes, we don't have suitable replacement for the old requests which can't yet move to the new handshake mechanism.

gavin-norman-sociomantic replied:
Mmm, isn't `handleSupportedCodes` it?

nemanja-boric-sociomantic replied:
No, because `handleSupportedCodes` can't tell the client if the actual status code was (previously) `GlobalStatusCode` or a request specific code - it returns false for all of them (it returns true only if the `RequestStarted` happens, which is never the case in the old request). In contrast, `handleGlobalStatusCodes` would return `true` if the status code was a global status code, but `false` if it was the request-specific status code.

Perhaps the deprecation path could be checking if the code is positive after casting it to `SupportedStatus` (signed byte), but that doesn't strike me as a nice solution.

gavin-norman-sociomantic replied:
Maybe I'm missing something...

A replacement for a deprecated method doesn't have to be a _direct_ replacement. It may require changes other than changing the name of the symbol used. The deprecation just indicates that something is changing and that all users of the code will have to migrate. Which is exactly the case we have here. Or?

nemanja-boric-sociomantic replied:
The problem is that I don't see a way to fix the old, but yet supported requests, without the change in the _protocol_ (this is in contrast to our usual deprecations that can be fixed without changing the on-wire protocol). That means that either we need to make a new version of the request, or we're stuck with the silly deprecation warning for which the proto users (so, not swarm users, but the users of the concrete proto implementations) can't do anything. Now, making a new version of the request just for this change, to get rid of this particular deprecation might be fine, but I was erring on the side of not having to do this.

nemanja-boric-sociomantic replied:
(I probably didn't use `erring` in a good context here).

gavin-norman-sociomantic replied:
Ah I see. You're right.
